petastorm/pytorch.py

```diff
@@ -420,5 +458,17 @@ def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.q:
+            # Close async shuffle thread
+            self.close_async_thread()
         self.reader.stop()
         self.reader.join()
+
+    def close_async_thread(self):
+        self.finished_event.set()
+        try:
+            # Free buffer to allow worker to retry
+            self.q.get_nowait()
```

Comment:
Not sure why this is needed.  Is this because the producer may be blocked trying to add an element to the queue?  Would it make sense to just clear the queue instead?

Reply:
It is for the boundary case, len(q) == q_size, and producer is waiting on q.put(), cannot finish join() if consumer doesn't consume something in the queue.